### PR TITLE
more aerosol metrics in e3sm_diags

### LINKED ISSUE #760: [Feature]: Add aerosol burden to lat-lon and other core sets
### Is your feature request related to a problem?

Global mean aerosol burden is included in the aerosol budget table. To add 2D aerosol burden in lat-lon sets would allow showing the global distribution as well as global metrics. It can also serve as back up if the aerosol budget table is not available. 

### Describe the solution you'd like

Including aerosol burden in standard lat-lon sets, and compare with obs/reanalysis if available. 

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mahf708**
## Description

Add preliminary code for aerosol-related diagnostics from the v3 atmosphere subgroup "Aerosol Working Group" led by Susannah Burrows. This work is partially from a collaborative effort by Naser Mahfouz, Johannes Mülmenstädt, and Susannah Burrows in 2023, and we picked a subset of diagnostics we think will be helpful to a wider audience when evaluating aerosols. This PR adds aerosol masses at different levels, aerosol integrated burdens, aerosol--cloud interaction metrics (e.g., cdnc, lwp), as well as aerosol effective radiative forcing. The PR also exposes all 2D "AeroCom" variables to the derived parameter list (thus enabling easier independent usage by users and/or extending e3sm_diags in future releases).

Caveats:
- The specs in cfg files are not final and will be edited based on feedback from users in upcoming releases.
- Note that the code added may result in non-fatal (harmless) runtime errors (which will appear in e3sm_diags logs) due to potential unavailability of all specified parameters in the h0 tape.

Closes #760 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings (⚠️ the added code will generate non-fatal runtime errors due to potentially unavailability of some diagnostics in the h0 tapes ⚠️)
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [x] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: e3sm_diags/derivations/acme.py [hunk 1]
```diff
@@ -2074,3 +2074,391 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+
+# Add burden vars to derived_variables
+for aero_burden_item in aero_burden_list:
+    derived_variables[aero_burden_item] = OrderedDict(
+        [((aero_burden_item,), aero_burden_fxn)]
+    )
+
+
+# Names of 2D mass slices of aerosol species
+# Also add 3D masses while at it (if available)
```
Comment (chengzhuzhang) [inline e3sm_diags/derivations/acme.py:2115]: I can't seem to find codes to support 3D masses..

**Review by chengzhuzhang — COMMENTED** [review]
Thank you @mahf708 for the PR. It looks great. This provides important update for supporting aerosol variables adjusted for upcoming v3 simulation. Also it is intriguing to see the use case with the model vs model run for ERF analysis, nice work! I only have a comment regarding supporting 3d mass for backward compatibility. I saw comments about it, but I think it hasn't been implemented?

I just tagged Susannah and Johannes for review. Appreciate this nice contribution from the aerosol working group!

File: e3sm_diags/derivations/acme.py [hunk 2]
```diff
@@ -2074,3 +2074,391 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+aero_mass_list = []
+for aero_name in ["dst", "mom", "pom", "so4", "soa", "ncl", "bc"]:
+    for aero_lev in ["srf", "200", "330", "500", "850", ""]:
+        aero_mass_list.append(f"Mass_{aero_name}_{aero_lev}")
```
[the remark this answers is not in the document]
Comment (chengzhuzhang) [reply e3sm_diags/derivations/acme.py:2119]: > @chengzhuzhang this is how I am taking the 3D mass variables (I needed to get rid of an extra `_`). Added a comment and correction.

Got it! Thank you.

File: e3sm_diags/derivations/acme.py (hunk 1 above)
Comment (chengzhuzhang) [reply e3sm_diags/derivations/acme.py:2115]: hey @mahf708, I should have be clearer, for 3D mass, I was going to suggest to have the function to [create column integrated](# https://github.com/E3SM-Project/e3sm_diags/blob/633b52c314325e605fe7f62687cc4d00e5a0a3d5/e3sm_diags/driver/aerosol_budget_driver.py#L29 ) value from 3D aerosol mass concentration in derived variable list under the entries of ABURDENXX, the entry in derived_variable would be something like.
```
"ABURDENBC": OrderedDict(
        [
            (("ABURDENBC",), aero_burden_fxn),
            # If ABURDENBC is not exist, detect if Mass_bc is available, and do column integral.
            (("Mass_bc","hyai", "hybi", "PS"), calc_column_integral),              
        ]
    ), 
```
This could be helpful when re-run e3sm_diags with simulations prior to v3.

Comment (chengzhuzhang) [reply e3sm_diags/derivations/acme.py:2115]: Sounds good. Other changes in this PR looks good to me!

**Comment by jmuelmen** [comment]
These diagnostics will be extremely useful for ACI analysis.  Great to have them in e3sm_diags!

**Review by jmuelmen — APPROVED** [review]

**Comment by susburrows** [comment]
Thanks, @mahf708, for implementing the changes in this PR and following through on the steps required to submit them to the e3sm_diags repository. These diagnostics are extremely valuable for understanding and tracking how the aerosol forcing and understanding the 3D distribution of aerosols change across model versions, while keeping the I/O burden manageable (e.g., by using 2D fields such as cross-sections of mass concentrations at several pressure levels).

Thanks for providing the opportunity for these contributions from the AWG to be included into e3sm_diags, @chengzhuzhang .

I think this looks good and am approving, based on examination of the test output.  If further changes are required in the future, I trust that those can be addressed in separate PRs.

**Review by susburrows — APPROVED** [review]

File: e3sm_diags/derivations/acme.py [hunk 3]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+    "ABURDENSEASALT",
+]
+
+
+def aero_burden_fxn(var):
+    """
+    Scale the aerosol burden by 1e6.
+
+    Parameters:
+        var (float): The input burden in kg/m2.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2097]:
```suggestion
    Parameters:
        var (cdms2.TransientVariable): The input burden in kg/m2.
```

I believe these are all `cdms2.TransientVariable` based on setting the `.units` attribute.

File: e3sm_diags/derivations/acme.py [hunk 4]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+
+def aero_mass_fxn(var):
+    """
+    Scale the given mass by 1e12.
+
+    Parameters:
+        var (float): The input mass in kg/kg.
+
+    Returns:
+        float: The aerosol mass concentration in 1e-12 kg/kg units.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2131]:
```suggestion
    Parameters:
        var (cdms2.TransientVariable): The input mass in kg/kg.

    Returns:
        cdms2.TransientVariable: The aerosol mass concentration in 1e-12 kg/kg units.
```

File: e3sm_diags/derivations/acme.py [hunk 5]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+def incldtop_cdnc(cdnc, lcc):
+    """
+    Return the in-cloud cloud droplet number concentration at cloud top.
+
+    Parameters:
+        cdnc (float): Cloud droplet number concentration in 1/m3.
+        lcc (float): Liquid cloud fraction.
+
+    Returns:
+        var (float): In-cloud cloud droplet number concentration at cloud top in 1/cm3.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2199]:
```suggestion
    Parameters:
        cdnc (cdms2.TransientVariable): Cloud droplet number concentration in 1/m3.
        lcc (cdms2.TransientVariable): Liquid cloud fraction.

    Returns:
        var (cdms2.TransientVariable): In-cloud cloud droplet number concentration at cloud top in 1/cm3.
```

File: e3sm_diags/derivations/acme.py [hunk 6]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+def cldtop_cdnc(cdnc):
+    """
+    Return the in-grid cloud droplet number concentration at cloud top.
+
+    Args:
+        cdnc (float): Cloud droplet number concentration in 1/m3.
+
+    Returns:
+        var (float): In-grid cloud droplet number concentration at cloud top in 1/cm3.
+    """
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2216]:
```suggestion
    Args:
        cdnc (cdms2.TransientVariable): Cloud droplet number concentration in 1/m3.

    Returns:
        var (cdms2.TransientVariable): In-grid cloud droplet number concentration at cloud top in 1/cm3.
    """
```

File: e3sm_diags/derivations/acme.py [hunk 7]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+def incldtop_icnc(icnc, icc):
+    """
+    Return the in-cloud ice crystal number concentration at cloud top.
+
+    Parameters:
+        icnc (float): ice crystal number concentration in 1/m3.
+        icc (float): ice cloud fraction.
+
+    Returns:
+        var (float): In-cloud cloud droplet number concentration at cloud top in 1/cm3.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2232]:
```suggestion
    Parameters:
        icnc (cdms2.TransientVariable): ice crystal number concentration in 1/m3.
        icc (cdms2.TransientVariable): ice cloud fraction.

    Returns:
        var (cdms2.TransientVariable): In-cloud cloud droplet number concentration at cloud top in 1/cm3.
```

File: e3sm_diags/derivations/acme.py [hunk 8]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+
+def cldtop_icnc(icnc):
+    """
+    Return the in-grid ice crystal number concentration at cloud top.
+
+    Args:
+        icnc (float): Cloud crystal number concentration in 1/m3.
+
+    Returns:
+        var (float): In-grid ice crystal number concentration at cloud top in 1/cm3.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2248]:
```suggestion
    Args:
        icnc (cdms2.TransientVariable): Cloud crystal number concentration in 1/m3.

    Returns:
        var (cdms2.TransientVariable): In-grid ice crystal number concentration at cloud top in 1/cm3.
```

File: e3sm_diags/derivations/acme.py [hunk 9]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+def incld_lwp(lwp, lcc):
+    """
+    Return the in-cloud liquid water path (LWP).
+
+    Parameters:
+        lwp (float): Liquid water path in kg/m2.
+        lcc (float): Liquid cloud fraction.
+
+    Returns:
+        float: In-cloud liquid water path in g/cm3.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2265]:
```suggestion
    Parameters:
        lwp (cdms2.TransientVariable): Liquid water path in kg/m2.
        lcc (cdms2.TransientVariable): Liquid cloud fraction.

    Returns:
        cdms2.TransientVariable: In-cloud liquid water path in g/cm3.
```

File: e3sm_diags/derivations/acme.py [hunk 10]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+
+def cld_lwp(lwp):
+    """
+    Return the grid-mean-cloud LWP in g/cm3.
+
+    Parameters:
+    lwp (float): Liquid Water Path (LWP) value.
+
+    Returns:
+    float: Grid-mean-cloud LWP in g/cm3.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2281]:
```suggestion
    Parameters:
		lwp (cdms2.TransientVariable): Liquid Water Path (LWP) value.

    Returns:
 	   cdms2.TransientVariable: Grid-mean-cloud LWP in g/cm3.
```

File: e3sm_diags/derivations/acme.py [hunk 11]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+def incld_iwp(iwp, icc):
+    """
+    Return the in-cloud ice water path (IWP).
+
+    Parameters:
+        iwp (float): Ice water path in kg/m2.
+        icc (float): Ice cloud fraction.
+
+    Returns:
+        float: In-cloud IWP in g/cm3.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2298]:
```suggestion
    Parameters:
        iwp (cdms2.TransientVariable): Ice water path in kg/m2.
        icc (cdms2.TransientVariable): Ice cloud fraction.

    Returns:
        cdms2.TransientVariable: In-cloud IWP in g/cm3.
```

File: e3sm_diags/derivations/acme.py [hunk 12]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+
+def cld_iwp(iwp):
+    """
+    Return the in-grid ice water path (IWP).
+
+    Parameters:
+        iwp (float): Ice water path in kg/m2.
+
+    Returns:
+        float: In-grid IWP in g/cm3.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2314]:
```suggestion
    Parameters:
        iwp (cdms2.TransientVariable): Ice water path in kg/m2.

    Returns:
        cdms2.TransientVariable: In-grid IWP in g/cm3.
```

File: e3sm_diags/derivations/acme.py [hunk 13]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+def erf_tot(fsnt, flnt):
+    """
+    Calculate the total effective radiative forcing (ERFtot).
+
+    Args:
+        fsnt (float): The incoming solar radiation at the top of the atmosphere.
+        flnt (float): The outgoing longwave radiation at the top of the atmosphere.
+
+    Returns:
+        var (float): The ERFtot which represents the total effect of radiative forcing.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2346]:
```suggestion
    Args:
        fsnt (cdms2.TransientVariable): The incoming solar radiation at the top of the atmosphere.
        flnt (cdms2.TransientVariable): The outgoing longwave radiation at the top of the atmosphere.

    Returns:
        var (cdms2.TransientVariable): The ERFtot which represents the total effect of radiative forcing.
```

File: e3sm_diags/derivations/acme.py [hunk 14]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+    Calculate aerosol--radiation interactions (ARI) part of effective radiative forcing (ERF).
+
+    Parameters:
+        fsnt (float): Net solar flux at the top of the atmosphere.
+        flnt (float): Net longwave flux at the top of the atmosphere.
+        fsnt_d1 (float): fsnt without aerosols.
+        flnt_d1 (float): flnt without aerosols.
+
+    Returns:
+        var (float): ERFari (aka, direct effect) in W/m2.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2367]:
```suggestion
    Parameters:
        fsnt (cdms2.TransientVariable): Net solar flux at the top of the atmosphere.
        flnt (cdms2.TransientVariable): Net longwave flux at the top of the atmosphere.
        fsnt_d1 (cdms2.TransientVariable): fsnt without aerosols.
        flnt_d1 (cdms2.TransientVariable): flnt without aerosols.

    Returns:
        var (cdms2.TransientVariable): ERFari (aka, direct effect) in W/m2.
```

File: e3sm_diags/derivations/acme.py [hunk 15]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+    Calculate aerosol--cloud interactions (ACI) part of effectie radiative forcing (ERF)
+
+    Parameters:
+        fsnt_d1 (float): Downward shortwave radiation at the top of the atmosphere without aerosols.
+        flnt_d1 (float): Upward longwave radiation at the top of the atmosphere without aerosols.
+        fsntc_d1 (float): fsnt_d1 without clouds.
+        flntc_d1 (float): flnt_d1 without clouds.
+
+    Returns:
+        var (float): ERFaci (aka, indirect effect) in W/m2.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2388]:
```suggestion
    Parameters:
        fsnt_d1 (cdms2.TransientVariable): Downward shortwave radiation at the top of the atmosphere without aerosols.
        flnt_d1 (cdms2.TransientVariable): Upward longwave radiation at the top of the atmosphere without aerosols.
        fsntc_d1 (cdms2.TransientVariable): fsnt_d1 without clouds.
        flntc_d1 (cdms2.TransientVariable): flnt_d1 without clouds.

    Returns:
        var (cdms2.TransientVariable): ERFaci (aka, indirect effect) in W/m2.
```

File: e3sm_diags/derivations/acme.py [hunk 16]
```diff
@@ -2074,3 +2074,392 @@ def cosp_histogram_standardize(cld: "FileVariable"):
...
+    Calculate the residual effect (RES) part of effective radiative forcin g.
+
+    Parameters:
+        fsntc_d1 (float): Downward solar radiation at the top of the atmosphere
+                          with neither clouds nor aerosols.
+        flntc_d1 (float): Upward longwave radiation at the top of the atmosphere
+                          with neither clouds nor aerosols.
+
+    Returns:
+        var (float): ERFres (aka, surface effect) in W/m2.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/acme.py:2409]:
```suggestion
    Parameters:
        fsntc_d1 (cdms2.TransientVariable): Downward solar radiation at the top of the atmosphere
                          with neither clouds nor aerosols.
        flntc_d1 (cdms2.TransientVariable): Upward longwave radiation at the top of the atmosphere
                          with neither clouds nor aerosols.

    Returns:
        var (cdms2.TransientVariable): ERFres (aka, surface effect) in W/m2.
```

**Review by tomvothecoder — APPROVED** [review]
I just had some minor docstring updates. Other than that it looks good to me.

---
On a side-note, I started refactoring `acme.py` in another [branch](https://github.com/E3SM-Project/e3sm_diags/tree/cdat-migration-fy24/e3sm_diags/derivations). I split it into `derivations.py`, `formulas.py`, and `utils.py`. These modules operate on Xarray objects instead of cdms2 objects. Also the `derived_variables` dictionary will eventually be cleaned up because it is massive and hard to follow.

**Comment by tomvothecoder** [comment]
> This is great! I intentionally put my additions at the end because I knew this was going to be refactored. I also think we may end up revisiting my haphazard implementation soon. Once, you're ready with your branch, I can push my additions to it, so that you don't have to bother rewriting this code ;)
>
>Also, great to have this done via xarray (I know xarray, but I know nothing about cmds2). 

Awesome! Whenever you're ready, you can open a PR that includes the Xarray version of the code. You just need to `cdat-migration-fy24` as the base branch.

This PR looks good to me.
